README.markdown
Botonaro
===

Chatbot que interage com material do Bolsodata do 
[Metamemo](https://metamemo.info/).  

Instruções
---

Atualmente a única interface é um bot de telegram usando uma biblioteca 
externa apropriada para esta finalidade.  

Crie um arquivo `.env` com um par chave = valor:  

    TELEGRAM_TOKEN = 123456:ABC-DEF1234ghIkl-zyx57W2v1u123ew11  

Para obter um token de Bot de Telegram, fale com o 
[@BotFather](https://t.me/botfather).  

```sh
$ python -m venv venv  
$ source venv/bin/activate # No Windows, Power Shell: .\venv\Scripts\activate  
(venv) $ pip install -e git+https://github.com/iuriguilherme/botonaro.git@stable#egg=botonaro  
(venv) $ pip install -e git+https://github.com/iuriguilherme/iacecil.git@stable#egg=iacecil  
(venv) $ python -m botonaro  
```

Para usar a funcionalidade de log e feedback em grupos do telegram, adicionar 
ao arquivo `.env`:  

```
ADMIN_CHATS = [1]
INFO_CHAT = -1
DEBUG_CHAT = -1
FEEDBACK_CHAT = -1
```

Substituindo 1s e -1s com os ids de usuário ou grupo pra quem o bot deve 
mandar mensagens de erro e etc.  

Para usar a funcionalidade principal de busca no metamemo, adicionar ao 
arquivo `.env`:  

```
BASE_URL = http://metamemo.info
LIST_ROUTE = /lista
ITEM_ROUTE = /metamemo
SOURCES_BUSCA = Facebook,Twitter,Youtube,Instagram,Telegram,Blog
START_DATE_BUSCA = 2022-01-01
CHANCE = 30
```

Adaptar **SOURCES_BUSCA** para o tipo de busca pertinente.  

Para usar a funcionalidade de integração com ChatGPT, adicionar ao 
arquivo `.env`:  

    OPENAI_API_KEY = sk-1234567890abcdef  

Para obter uma chave de API, acesse 
<https://platform.openai.com/account/api-keys>  

Roadmap
---

### [Pre-alpha v0.2](https://github.com/iuriguilherme/botonaro/releases/tag/pre-alpha)

Protótipo para apresentação para o Bolsodata  

### [Alpha v0.3](https://github.com/iuriguilherme/botonaro/releases/tag/alpha)

Chatbot de telegram, extensão de personalidade de 
[ia.cecil](https://github.com/iuriguilherme/iacecil)  

1. Lê quatro tipos de mensagem:
  1. Aquelas que contém o gatilho "fala sobre"
  1. Aquelas que começam com o comando "/sobre"
  1. Todas aquelas que foram enviadas para um chat particular
  1. 1:30 chance das que forem enviadas para um grupo
1. Usa o texto da mensagem para buscar na fonte de frases
1. Responde a mensagem com a frase, incluindo a referência

### [Alpha v0.4](https://github.com/iuriguilherme/botonaro/releases/tag/v0.4.0)

_(em desenvolvimento)_  

1. Funcionalidadas herdadas da versão 0.3
1. Registra reações a mensagens
1. Interface web com gráficos e estatísticas de reações
1. A cada mensagem, uma chance de responder usando ChatGPT

Licença
---

Copyright 2022 Iuri Guilherme <https://iuri.neocities.org/>  

[Creative Commons 4.0 Attribution Share Alike](LICENSE.markdown)  


CHANGES.txt
v0.4.3.1, 2022-12-14 -- Site explicativo


src/instance/personalidades/custom.py
"""
Personalidade ia.cecil: Botonaro

Copyright 2022 Iuri Guilherme <https://iuri.neocities.org/>  

Creative Commons 4.0 Attribution Share Alike  
"""
import logging
logger = logging.getLogger(__name__)

import aiohttp
import bs4
import datetime
import locale
import os
import random
import typing
from aiogram import (
    Dispatcher,
    filters,
    types,
)
from aiogram.utils import markdown
## FIXME engambelada enquanto ia.cecil não migra pra aiogram 3
from aiogram.types import InputFile as URLInputFile
from iacecil.controllers.aiogram_bot.callbacks import (
    command_callback,
    message_callback,
    error_callback,
)
from iacecil.controllers.util import (
    dice_high,
    dice_low,
)

try:
    locale.setlocale(locale.LC_TIME, 'pt_BR.UTF-8')
except Exception as e:
    logger.debug("Locale brasileira não instalada no sistema")
    logger.exception(e)

class ZeroResultsException(Exception):
    pass

class ItemMixin(object):
    """Modelo para item individual de busca"""
    link: typing.Union[str, None] = None
    fonte: str = "Internet"
    metamemo: str = "Alguém"
    texto: str = "Nada"
    video: typing.Union[str, None] = None
    imagem: typing.Union[str, None] = None
    apoios: str = "0"
    comentarios: str = "0"
    data: datetime.datetime = datetime.datetime.min
    def __repr__(self):
        return f"""<ItemMixin(\
link: {str(self.link)}, \
fonte: {self.fonte}, \
metamemo: {self.metamemo}, \
texto: {self.texto}, \
video: {str(self.video)}, \
imagem: {str(self.imagem)}, \
apoios: {self.apoios}, \
comentarios: {self.comentarios}, \
data: {str(data)}\
>"""

class ResultMixin(object):
    """Modelo para resultado da busca"""
    link: str
    titulo: str
    fonte: str
    data: datetime.datetime
    metamemo: str
    url: str
    midia: list[str]
    def __repr__(self):
        return f"""<ResultMixin(\
link: {self.link}, \
titulo: {self.titulo}, \
fonte: {self.fonte}, \
data: {str(self.data)}, \
metamemo: {self.metamemo}, \
url: {self.url}, \
midia: {",".join(self.midia)}\
>"""

async def html_to_bytes(url: str) -> typing.Union[bytes, None]:
    """Performs async HTTP GET on URL and returns bytes for parsing"""
    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(url) as response:
                return await response.read()
    except Exception as e:
        logger.exception(e)
        return None

async def html_to_json(url: str) -> typing.Union[str, None]:
    """Performs async HTTP GET on URL and returns JSON string"""
    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(url) as response:
                return await response.json()
    except Exception as e:
        logger.exception(e)
        return None

async def bytes_to_soup(html: bytes) -> typing.Union[bs4.BeautifulSoup, None]:
    """Parse bytes to beautiful soup"""
    try:
        return bs4.BeautifulSoup(html, "lxml")
    except bs4.FeatureNotFound:
        logger.debug("python-lxml not installed, falling back to html.parser")
        return bs4.BeautifulSoup(html, "html.parser")
    except Exception as e:
        logger.exception(e)
        return None

async def parse_item(rows: bs4.element.ResultSet) -> ItemMixin:
    """Parse HTML into Python Object"""
    try:
        item: ItemMixin = ItemMixin()
        try:
            item.link: str = rows[0].find('a', {'title': "Link original"}
                ).get('href')
        except Exception as e:
            logger.debug("Item provavelmente não tem link")
            logger.exception(e)
        try:
            item.fonte: str = rows[0].find('span').get('class')[0].split('-'
                )[1].strip('1').capitalize()
        except Exception as e:
            logger.debug("Item provavelmente não tem fonte")
            logger.exception(e)
        try:
            item.metamemo: str = rows[0].find('b').text
        except Exception as e:
            logger.debug("Item provavelmente não tem metamemo")
            logger.exception(e)
        try:
            item.texto: str = rows[1].find('p').text
        except Exception as e:
            logger.debug("Item provavelmente não tem texto")
            logger.exception(e)
        try:
            item.video: str = rows[2].find('video', 'metavideo').find('source'
                ).get('src')
        except Exception as e:
            logger.debug("Item provavelmente não tem vídeo")
            logger.exception(e)
        try:
            item.imagem: str = rows[2].find('img').get('src')
        except Exception as e:
            logger.debug("Item provavelmente não tem imagem")
            logger.exception(e)
        try:
            item.apoios: str = rows[3].find_all('b')[1].text
        except Exception as e:
            logger.debug("Item provavelmente não tem curtidas")
            logger.exception(e)
        try:
            item.comentarios: str = rows[3].find_all('b')[2].text
        except Exception as e:
            logger.debug("Item provavelmente não tem comentários")
            logger.exception(e)
        try:
            item.data: datetime.datetime = datetime.datetime.strptime(
                " ".join([
                    b.text.strip('\xa0 ') \
                    for b in \
                    rows[3].find('div', 'data-tempo').find_all('b')
                ]), "%H:%M %d/%m/%Y"
            ) # item.data
        except Exception as e:
            logger.debug("Item provavelmente não tem data")
            logger.exception(e)
        return item
    except Exception as e:
        logger.exception(e)
        return None

async def parse_result(result_item: bs4.element.Tag) -> ResultMixin:
    """Parse HTML into Python Object"""
    try:
        result: ResultMixin = ResultMixin()
        ## Nota do desenvolvedor: eu ODEIO isto. Isso é a coisa mais feia, 
        ## chata e imbecil que existe no mundo da programação. É muito fácil 
        ## editar o HTML que é gerado e quebrar totalmente o modelo que 
        ## esta função cria, me forçando a abrir o navegador, o terminal do 
        ## python e a IDE pra reescrever tudo de novo, do zero. Este aqui é o 
        ## melhor exemplo de porque é que uma coisa que "está funcionando" não 
        ## significa sob hipótese alguma que daqui a alguns segundos ainda vai 
        ## estar "funcionando". Isso aqui não é programar, isso aqui é brincar 
        ## de ficar fazendo gambiarra. No pior sentido. Isso aqui NÃO presta, 
        ## NÃO funciona, NÃO resolve, NÃO soluciona. Só causa problemas e toma 
        ## tempo desnecessariamente. O nome das tags mudou pelo menos uma vez 
        ## durante a fase de testes desse conjunto de subrotinas. 
        ## RESTful API or GTFO. </rant>
        columns: bs4.element.ResultSet = result_item.find_all('div', 's1')
        result.link: str = columns[0].find('a', 'icon-eye').get('href')
        result.titulo: str = result_item.find('div', 's5').text
        result.fonte: str = columns[1].find('i').get('class')[1].split('-'
            )[1].strip('1').capitalize()
        result.data: datetime.datetime = datetime.datetime.strptime(
            " ".join([columns[2].text, columns[3].text]), "%d/%m/%Y %H:%M")
        result.metamemo: str = columns[4].text
        result.url: str = columns[5].find('a').get('href')
        result.midia: list[str] = [a.get('href') for a in \
            columns[6].find_all('a')]
        return result
    except Exception as e:
        logger.exception(e)
        return None

async def busca_item(link: str) -> ItemMixin:
    """Retorna os dados de um dos itens de uma busca a partir de link"""
    try:
        fonte_url: str = os.environ.get("BASE_URL", "http://example.com")
        url: str = f"{fonte_url}/" + f"{link}"
        logger.debug(f"Buscando item em {url}")
        soup: BeautifulSoup = await bytes_to_soup(await html_to_bytes(url))
        results: bs4.element.ResultSet = soup.find('div', 'memoitem-container',
            ).find_all('div', 'row')
        return await parse_item(results)
    except Exception as e:
        logger.exception(e)
        item: ItemMixin = ItemMixin()
        item.texto: str = ""
        return item

async def busca_frase(
    palavras: list[str],
) -> typing.Union[list, list[ResultMixin]]:
    """Retorna o resultado de uma busca em BASE_URL"""
    try:
        fonte_url: str = "/".join([
            os.environ.get("BASE_URL", "http://example.com"),
            os.environ.get("LIST_ROUTE", "/"),
        ])
        sources: list = os.environ.get("SOURCES_BUSCA").split(',')
        start_date: str = os.environ.get("START_DATE_BUSCA")
        end_date: str = datetime.datetime.today().strftime("%Y-%m-%d")
        url: str = f"{fonte_url}/?" \
            + "&".join([f"source={source}" for source in sources]) \
            + f"&content={'%20'.join(palavras)}" \
            + f"&start_date={start_date}" \
            + f"&end_date={end_date}"
        logger.debug(f"Buscando em {url}")
        soup: BeautifulSoup = await bytes_to_soup(await html_to_bytes(url))
        results: bs4.element.ResultSet = soup.find('div', 'search-results')
        results: bs4.element.ResultSet = results.find_all('div', 'result-item')
        retornos: list = [await parse_result(result) for result in results]
        logger.debug(f"Encontrados {len(retornos)} resultados")
        return retornos
    except Exception as e:
        logger.exception(e)
        return []

async def add_instance_handlers(dispatcher: Dispatcher) -> None:
    """Registra handlers para aiogram.Dispatcher, lida com Telegram"""
    try:
        @dispatcher.message_handler(commands = ['start', 'help', 'info'])
        async def start_callback(message: types.Message) -> None:
            """Resposta específica para comando /start"""
            await message.reply("""Pode mandar qualquer termo ou frase que \
eu vou tentar achar na internet o que é que a minha família disse sobre isso \
aí ta ok""")
        async def busca_callback(
            message: types.Message,
            palavras: str,
            descriptions: list,
        ) -> typing.Union[types.Message, None]:
            """Busca frase de acordo com palavras chave"""
            try:
                frases: list[ResultMixin] = [
                    frase for frase in \
                    await busca_frase(palavras) \
                    if hasattr(frase, 'link')
                ]
                if len(frases) < 1:
                    raise ZeroResultsException("Nenhum resultado")
                else:
                    item: ItemMixin = await busca_item(random.choice(
                        frases).link)
                    formato_data: str = "em %d/%m/%Y às %H:%M"
                    if locale.getlocale()[0] == "pt_BR":
                        formato_data: str = "%A, %d/%m/%Y às %H:%M"
                    captions: list = [f"""Publicado por {item.metamemo} \
{item.data.strftime(formato_data)}."""]
                    if int(item.apoios) + int(item.comentarios) > 0:
                        captions.append(f"{item.apoios} " + u"\U0001f44d" + \
                            f"{item.comentarios} " + u"\U0001f4ac")
                    if item.link not in [None, '', ' ']:
                        captions.append(f"Link: {item.link}")
                    caption: str = markdown.spoiler("\n".join(captions))
                    texto: str = "\n".join([markdown.escape_md(item.texto),
                        "", caption])
                    if len(texto) < 24+1e3:
                        caption: str = texto
                if item.video is not None:
                    try:
                        return await message.reply_video(
                            video = URLInputFile.from_url(item.video),
                            caption = caption,
                            parse_mode = "MarkdownV2",
                            disable_notification = True,
                            # ~ disable_web_page_preview = True,
                            allow_sending_without_reply = True,
                        )
                    except Exception as e2:
                        logger.exception(e2)
                        await error_callback("Erro tentando mandar vídeo",
                            message, e2, ['exception'] + descriptions)
                elif item.imagem is not None:
                    try:
                        return await message.reply_photo(
                            photo = URLInputFile.from_url(item.imagem),
                            caption = caption,
                            parse_mode = "MarkdownV2",
                            disable_notification = True,
                            # ~ disable_web_page_preview = True,
                            allow_sending_without_reply = True,
                        )
                    except Exception as e2:
                        logger.exception(e2)
                        await error_callback("Erro tentando mandar imagem",
                            message, e2, ['exception'] + descriptions)
                return await message.reply(
                    text = texto,
                    parse_mode = "MarkdownV2",
                    disable_notification = True,
                    disable_web_page_preview = True,
                    allow_sending_without_reply = True,
                )
            except Exception as e1:
                logger.exception(e1)
                raise
        async def busca_quieta(message: types.Message) -> None:
            """
            Busca através de palavras na mensagem, não responde se não achar 
            nada
            """
            try:
                descriptions: list = [
                    'botonaro',
                    'buscamemo',
                    'buscanatural',
                    dispatcher.config.personalidade,
                    message.chat.type,
                ] # descriptions
                await message_callback(message, descriptions)
                try:
                    command: types.Message = await busca_callback(
                        message,
                        [
                            termo \
                            for termo in \
                            message.text.split(' ') \
                            if termo not in \
                            ["fala", "Fala", "sobre", "start"]
                        ],
                        descriptions,
                    ) # command
                except ZeroResultsException:
                    logger.debug("sem resultados")
                else:
                    await command_callback(command, descriptions)
            except Exception as e1:
                logger.exception(e1)
                await error_callback("Erro buscando frase", message,
                    e1, ['exception'] + descriptions)
        async def busca_responde(message: types.Message) -> None:
            """
            Busca através de palavras na mensagem e informa se não achar nada
            """
            try:
                descriptions: list = [
                    'botonaro',
                    'buscamemo',
                    'buscanatural',
                    dispatcher.config.personalidade,
                    message.chat.type,
                ] # descriptions
                await message_callback(message, descriptions)
                try:
                    command: types.Message = await busca_callback(
                        message,
                        [
                            termo \
                            for termo in \
                            message.text.split(' ') \
                            if termo not in \
                            ["fala", "Fala", "sobre", "start"]
                        ],
                        descriptions,
                    ) # command
                except ZeroResultsException:
                    command: types.Message = await message.reply(
                        text = """não me recordo de nada no tocante a essa \
qüestão aí talquei""",
                        disable_notification = True,
                        allow_sending_without_reply = True,
                    ) # command
                await command_callback(command, descriptions)
            except Exception as e1:
                logger.exception(e1)
                await error_callback("Erro buscando frase", message,
                    e1, ['exception'] + descriptions)
        @dispatcher.message_handler(commands = ['sobre', 'm'])
        async def busca_comando_callback(message: types.Message) -> None:
            """Busca através dos argumentos do comando"""
            try:
                descriptions: list = [
                    'botonaro',
                    'buscamemo',
                    'buscacomando',
                    dispatcher.config.personalidade,
                    message.chat.type,
                ] # descriptions
                await message_callback(message, descriptions)
                try:
                    command: types.Message = await busca_callback(
                        message,
                        message.get_args().split(' '),
                        descriptions,
                    ) # command
                except ZeroResultsException:
                    command: types.Message = await message.reply(
                        text = """não me recordo de nada no tocante a essa \
qüestão aí talquei""",
                        disable_notification = True,
                        allow_sending_without_reply = True,
                    ) # command
                await command_callback(command, descriptions)
            except Exception as e1:
                logger.exception(e1)
                await error_callback("Erro buscando frase", message, e1, 
                    ['exception'] + descriptions)
        @dispatcher.message_handler(
            filters.Regexp(r'\bfala sobre\b'),
            content_types = types.ContentTypes.TEXT,
        )
        async def busca_natural_callback(message: types.Message) -> None:
            await busca_responde(message)
        @dispatcher.message_handler(
            filters.ChatTypeFilter(types.ChatType.PRIVATE),
            content_types = types.ContentTypes.TEXT,
        )
        async def busca_private_callback(message: types.Message) -> None:
            await busca_responde(message)
        @dispatcher.message_handler(
            content_types = types.ContentTypes.TEXT,
            state = "*",
        )
        async def chance_busca_callback(message: types.Message) -> None:
            """Responde em uma chance aleatória"""
            try:
                if await dice_low(int(os.environ.get("CHANCE", 30))):
                    await busca_quieta(message)
                else:
                    logger.debug("sem resposta")
            except Exception as e1:
                logger.exception(e1)
                await error_callback("Erro buscando frase", message,
                    e1, ['exception'])
    except Exception as e:
        logger.error("Não consegui registrar os handlers de busca")
        logger.exception(e)
